nova/scheduler/weights/resize_same_shard.py

```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2022 SAP
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+Prefer resize to same shard weigher. Resizing an instance gives the current
+aggregate a higher priority.
+
+This weigher ignores any instances which are:
+ * new instances (or unshelving)
+ * not resizing (a.k.a. migrating or rebuilding)
+"""
+from oslo_log import log as logging
+
+import nova.conf
+from nova.scheduler import utils
+from nova.scheduler import weights
+
+LOG = logging.getLogger(__name__)
+
+CONF = nova.conf.CONF
+
+
+class PreferSameShardOnResizeWeigher(weights.BaseHostWeigher):
+    minval = 0
+    _SHARD_PREFIX = 'vc-'
+
+    def weight_multiplier(self):
+        """Override the weight multiplier."""
+        return (CONF.filter_scheduler.
+            prefer_same_shard_resize_weight_multiplier)
+
+    def _weigh_object(self, host_state, request_spec):
+        """Return 1 for about-to-be-resized instances where the shard is the
+        instance's current shard. Return 0 otherwise.
+        """
+        if (request_spec.instance_uuid not in host_state.instances
+                or not utils.request_is_resize(request_spec)):
+            return 0.0
+
+        host_shard_aggrs = [aggr for aggr in host_state.aggregates
+                            if aggr.name.startswith(self._SHARD_PREFIX)]
+        if not host_shard_aggrs:
+            LOG.warning('%(host_state)s is not in an aggregate starting with '
+                        '%(shard_prefix)s.',
+                        {'host_state': host_state,
+                         'shard_prefix': self._SHARD_PREFIX})
+            return 0.0
+
+        instance = host_state.instances[request_spec.instance_uuid]
+        instance_host = instance.get("host")
+        if instance_host is None:
+            instance_host = request_spec.get_scheduler_hint('source_host')
+        host_shard_hosts = set(host
+                               for aggr in host_shard_aggrs
+                               for host in aggr.hosts)
```

Comment:
There really should be only one VC aggregate for a host (relates also to lines 51-58). I'd prefer an update of the code to look like #306 where we also warn if there a multiple. It might make sense to push retrieving the vc-aggregate for a host into a helper-function as I think we use that somewhere else, too - probably in the `ShardFilter`. Moving that code into a helper-function should be its own commit, though.

Reply:
I will create separate PR(after this merged) for it considering all places.